Return error if provided SRV record can't be resolved

victor-perov opened the pull request:
Replace panicing in randomBroker() with returning an error
    
    This func is used internally in the only place: connectToRandomBroker(),
    which already returns error so we can replace panic() safely.

Add panic to randomBroker() if provided SRV record can't be resolved

    `randomBroker()` accepts a name to be resolved for further client
    calls. It could be a single broker name or an SRV record. If it's
    an SRV record that for some reason resolves into an empty list of
    hosts, `randomBroker()` will panic.
    If the provided name is not an SRV, it will be used as is.

victor-perov committed:
Add panic to randomBroker() if provided SRV record can't be resolved

`randomBroker()` accepts a name to be resolved for further client
calls. It could be a single broker name, or an SRV record. If it's
an SRV record that for some reasons resolves into an empty list of
hosts, `randomBroker()` will panic.
If provided name is not an SRV, it will be used as is.

kvap commented on client/client.go:
```diff
@@ -64,17 +68,20 @@ func forgetSrv(name string) {
+		if errors.Is(err, errNotAnSRV) {
+			return name
+		}
+		panic(err)
```
Returning an error and handling it up the chain would be better.

victor-perov replied:
this is what I decided to do when I woke up today.

victor-perov committed:
Replace panicing in randomBroker() with returning an error

This func is used internally in the only place: connectToRandomBroker(),
which already returns error so we can replace panic() safely.

kvap commented on client/client.go:
```diff
@@ -38,20 +41,21 @@ func lookupSrv(name string) ([]string, error) {
+		if err != nil || len(hosts) == 0 {
+			continue
 		}
 		host := net.JoinHostPort(hosts[0], strconv.Itoa(int(srv.Port)))
```
Why do we only append one of the addresses of the host?
Why not put the `srv.Target` into the list? The Dial function could then resolve it again and use any of the addresses.

victor-perov replied:
I don't think I got your question. Would mind elaborating? How is that one related to the PR?

victor-perov replied:
I see. I think the logic makes sense: an SRV record could contain a slice of targets, and sometimes the host could be broken and unresolvable. Because we have a slice of hosts, we can filter out those broken ones and continue issuing requests correctly, instead of failing during dialling to the host. Also, we do need to obey the srv.Port that is advertized by an SRV.

kvap commented on client/client.go:
```diff
@@ -38,20 +41,21 @@ func lookupSrv(name string) ([]string, error) {
 	_, srvs, err := net.LookupSRV("", "", name)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %s", errNotAnSRV, err)
```
Usually the "cause" is wrapped and goes after the colon. Did you mean `%v: %w`?

kvap replied:
An error here does not always mean it is "not an SRV".

victor-perov replied:
The error means we don't need to apply SRV resolving logic here because that's not an SRV for sure.

kvap replied:
I see how it is intended by you, but it is the "for sure" part that I disagree with.

kvap replied:
The failure can be caused by "not an SRV", or by a network timeout during the DNS communications, or a buggy DNS server sending malformed response, or something else.

victor-perov replied:
I see why are you arguing.

I didn't mean to say that  `net.LookupSRV()` is a method to check is it an SRV or not.
What I've tried to say, is that we have SRV and not SRV logic. My PR continues the best-effort way to resolve the record.
And `errNotAnSRV` helps to switch logic (further resolving or using as-is).

kvap commented:
LGTM :+1: , just a couple of small comments.
